source/common/network/listener_impl.cc

```diff
@@ -27,7 +27,11 @@ void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr*
                                   int remote_addr_len, void* arg) {
   ListenerImpl* listener = static_cast<ListenerImpl*>(arg);
 
-  Address::InstanceConstSharedPtr final_local_address = listener->socket_.localAddress();
+  // Get the local address from the new socket.  This is different from the listener address, if
+  // the listener is listening on the all hosts address (e.g., 0.0.0.0 for IPv4).
+  Address::InstanceConstSharedPtr local_address = Address::addressFromFd(fd);
```

Comment:
In addition to the address from fd, down the line I see a need to have other sources for original destination address and other upstream metadata information than the SO_ORIGINAL_DST. Given this it would be good to package this socket option to a mockable interface as well already now.